# Fibres in the left ventricle

#### A dataset from the September 2015 Data Input and Manipulation Workshop in Bloemfontein South Africa
Data descrition from Kevin Sack:

These are fibres in the left ventricle - their colour is indicative of the angle the fibres make with the circumferential direction of the heart. (i.e blue points downwards, towards the apex/tip of the heart, and red point upwards, towards the aorta). The complex fibre structure makes for an interesting 3D description.
<img src = "images/fibres.png"/>

## Data Import

In [3]:
#Don't really need to use the astronomy package, but...
from astropy.table import Table, Column
from astropy.io import ascii

In [4]:
LineTable = ascii.read("Fibre_LV_export_TCarr_2015.txt")

The table consists of one line segment per line with an attached color: (x0,y0,z0,x1,y1,z1,(r,g,b)) 

In [5]:
LineTable

<Table masked=False length=5145>
  col1     col2     col3     col4     col5     col6        col7     
float64  float64  float64  float64  float64  float64    string112   
-------- -------- -------- -------- -------- -------- --------------
-28.6301 -13.6001 -17.6625 -28.6624 -14.3822 -17.0403 1.00 0.78 0.00
 -0.4986 -31.9882  13.4586   0.2171 -32.0938  14.1489 1.00 0.80 0.00
 18.2536  11.7219 -53.6276  18.0807  12.6981 -53.4965 0.70 1.00 0.30
 25.0258   7.8884 -50.1305  24.6911   8.7785 -50.4399 0.00 0.89 1.00
-24.9116  26.4442 -39.6386 -25.4163  25.8142 -40.2288 0.00 0.48 1.00
 -0.6536 -33.9619  12.0535   0.2392  -34.032  12.4985 0.83 1.00 0.17
 23.4381  18.0725 -38.1475  23.0643  18.9746 -37.9316 0.63 1.00 0.37
 -35.121   4.0726 -23.7402 -35.2193   3.1332 -24.0686 0.00 0.85 1.00
 20.4241   8.9226 -48.4303  20.8559   9.7365 -48.0416 1.00 0.63 0.00
-13.0956  16.3854 -55.8692 -14.0594   16.145  -55.754 0.74 1.00 0.26
     ...      ...      ...      ...      ...      ...            ...
-13.9439   -8.814 -66.9563  -13.258  -9.4473 -67.3147 0.00 0.76 1.00
-19.9652 -25.0949   9.2505  -19.557 -25.6222   9.9957 1.00 0.70 0.00
-37.1879  23.0216  14.2781 -37.7274  22.2196  14.0217 0.00 0.93 1.00
 28.6086  -3.1006   1.3785  28.9309  -2.5417   2.1426 1.00 0.61 0.00
 -3.0493  37.7408  13.5352  -4.0322  37.6846  13.7105 0.48 1.00 0.52
 -28.873  27.3652 -10.7713 -29.5521  26.6679 -11.0006 0.00 0.98 1.00
  -4.825  33.8798  14.3896  -5.4375  33.9113  15.1794 1.00 0.61 0.00
-31.1056  28.5563   -9.439 -31.6697  27.9453  -9.9945 0.00 0.52 1.00
  -35.75  21.7871  13.1683 -36.2883  20.9465  13.2281 0.33 1.00 0.67
  7.6789 -42.3796 -20.2235   8.5828 -42.1962 -20.6099 0.00 0.76 1.00

## DigitalSky Visualization

There are various ways to visualize sets of lines on DigitalSky
1. As a 3D model
2. As a [.vla](http://paulbourke.net/dataformats/vla/) file
3. As a partiview mesh object

Building a 3D model is cumbersome and makes the data difficult to modify, so we won't do that.

The .vla file is nice, however the format only allows you to provide intensities rather than individual colors to the line segments. This is an artifact of the fact that the format dates back to the days of monochrome vector displays.

We'll construct the visualization using the partiview style mesh. The mesh objects are rectangular grids with the vertices displaced. We'll draw each line segment as a a 1x2 rectangular mesh rendered as a wireframe, like this:

    mesh -c 2 -s wire {
      1 2
      -0.4986 -31.9882  13.4586 
      0.2171 -32.0938  14.1489 
    }

These mesh blocks can be chained together in the .speck file


In [12]:
writefile=open('Fibre_LV.speck',"w")
i=0
for Line in LineTable:
    writefile.write("mesh -c %d -s wire {\n" % i)
    writefile.write("1 2\n")
    writefile.write("%8.4f %8.4f %8.4f \n" % (Line[0],Line[1],Line[2]))
    writefile.write("%8.4f %8.4f %8.4f \n" % (Line[3],Line[4],Line[5]))
    writefile.write("}\n")
    i=i+1
writeFile.close

It turns out the colormap scheme we've tried below isn't working. DigitalSky seems to only support approximately 100  colors in the .cmap file. The better wat to have done it would be to apply a colormap directly with alpha_h. Requesting a new data file from Tim Carr

In [25]:
writefile=open('Fibre_LV.cmap',"w")
writeFile.write("%d\n" % len(LineTable))
for Line in LineTable:
    writefile.write("%s \n" % Line[6])
writeFile.close()